### ISSUE TITLE
Bug in writing a LongLong value using MemLongLong (Let property)

### ISSUE THREAD

**Issue by vgrstn** [cristianbuse/VBA-MemoryTools#2]
Hi Chistian,

I figured you can actually apply the implemented technology to get a LongLong into a Currency variable. In this way you don't need to divide by 1000 and - and least in theory - loose some digits. 

I did a similar implementation of the weakreference concepts

Let me know what you think. 

Cheers

Vincent

Public Property Let MemLongLong(ByVal memAddress As LongLong, ByVal newValue As LongLong)
    #If Mac Then
        CopyMemory ByVal memAddress, newValue, 8
    #Else
        'Cannot set Variant/LongLong ByRef so we use a Currency instead

        LinkMem m_remoteMemory, newValue, vbCurrency ' not BY_Ref
        DIM CopyLongLong As Currency
        CopyLongLong = m_remoteMemory.memValue

        LinkMem m_remoteMemory, memAddress, vbCurrency + VT_BYREF
        LetByRefCurr(m_remoteMemory.memValue) = CopyLongLong 
        m_remoteMemory.memValue = Empty
    #End If
End Property

**Comment by cristianbuse** [comment]
Hi Vincent,

Thanks for sending your code. Nice idea. I am glad that someone else is also looking into stuff like this.

Unfortunately, I am going to keep the current version only for speed considerations. Here is how I tested:

Drop this code into the ```LibMemory``` module:  
```VBA
#If Win64 Then
Sub DemoMemLongLongSpeed()
    Dim x1 As LongLong: x1 = 111111111111111^
    Dim x2 As LongLong
    Const iterations As Long = 10000000
    
    Dim i As Long
    Dim t As Double
    '
    t = Timer
    For i = 1 To iterations
        MemLongLong(VarPtr(x2)) = x1
    Next i
    Debug.Print "MemLongLong (Let) " & Format$(iterations, "#,##0") & " times in " & Round(Timer - t, 3) & " seconds"
    Debug.Assert x1 = x2
    '
    t = Timer
    For i = 1 To iterations
        MemLongLong2(VarPtr(x2)) = x1
    Next i
    Debug.Print "MemLongLong2 (Let) " & Format$(iterations, "#,##0") & " times in " & Round(Timer - t, 3) & " seconds"
    Debug.Assert x1 = x2
End Sub
#Else
#End If
```
Where  ```MemLongLong``` is the current implementation and ```MemLongLong2``` is your implementation. Here is the code for both, if it's easier:  
```VBA
Public Property Let MemLongLong(ByVal memAddress As LongLong, ByVal newValue As LongLong)
    #If Mac Then
        CopyMemory ByVal memAddress, newValue, 8
    #Else
        'Cannot set Variant/LongLong ByRef so we use a Currency instead
        Const currDivider As Currency = 10000
        LinkMem m_remoteMemory, memAddress, vbCurrency + VT_BYREF
        LetByRefCurr(m_remoteMemory.memValue) = CCur(newValue / currDivider)
        m_remoteMemory.memValue = Empty
    #End If
End Property

Public Property Let MemLongLong2(ByVal memAddress As LongLong, ByVal newValue As LongLong)
    #If Mac Then
        CopyMemory ByVal memAddress, newValue, 8
    #Else
        LinkMem m_remoteMemory, newValue, vbCurrency ' not BY_Ref
        Dim CopyLongLong As Currency
        CopyLongLong = m_remoteMemory.memValue
    
        LinkMem m_remoteMemory, memAddress, vbCurrency + VT_BYREF
        LetByRefCurr(m_remoteMemory.memValue) = CopyLongLong
        m_remoteMemory.memValue = Empty
    #End If
End Property
```

I got something like this:  
![demo](https://i.stack.imgur.com/wrng7.png)

The speed is very close considering I ran 10 million iterations but I will keep the faster one of course.

One last thing, dividing by 10000 does not loose any digits. That's because ```Currency``` is really a glorified ```LongLong```. The storage for ```Currency``` is exactly the same as any integer except VBA abstracts and "says" the last 4 digits are decimals. Of course the discussion would be different for a floating-point number like ```Double``` but here there is really no issue.

Thanks again!

BTW, I will push some new methods quite soon to this repository so it might be worth if you click Watch in the top-right corner of GitHub (next to Star) so that you get notified when I push.

Regards,
Cristian

**Comment by vgrstn** [comment]
Hello Christian I do appreciate the speed difference. However, I do think your "mathematical" route does have a issue.. I use the function to manipulate pointers and I just want to be sure nothing changes in the bit pattern when moving from a LongLong to a Currency (and vice versa). I thought about a small experiment in code. I came up with the code below..

Please run the code and let me know.

Cheers

Vincent

Sub test()

    Const vbLongLongMin As LongLong = -(2 ^ (8 * 8 - 1))
    Const vbLongLongMax As LongLong = (2 ^ (8 * 8 - 1)) - 1
    
    Dim ll As LongLong
    
    ll = vbLongLongMax
    
    Debug.Print vbLongLongMin
    Debug.Print vbLongLongMax
    
    Dim c As Currency
    Dim i As Long
    
    ' first do the negative value (works!)

    For i = 1 To 8
        ll = -(2 ^ (8 * i - 1))
        c = CCur(ll / 10000)
        Debug.Print i, ll, c
    Next

    ' now do the positive value (numbers doen't match and overflow at i = 7)

    For i = 1 To 8
        ll = (2 ^ (8 * i - 1)) - 1
        c = CCur(ll / 10000)
        Debug.Print i, ll, c
    Next

End Sub

**Comment by vgrstn** [comment]
Using Currency is not the issue. That is just a bit pattern like Long Long. Dividing by 10000 is an issue, at least for positive numbers.

**Comment by cristianbuse** [comment]
Hi Vincent,

Thanks for coming back on this. Indeed you are correct. At the upper limit of positive LongLong values, the mathematical approach (division by 10k ) fails by overflow and also precision is lost changing the last digit.

I do not know how often you would have pointers at the upper positive limit but regardless of that, your solution is clean. Well done! I've already pushed a commit that fixes the issue. Also, I've removed the calls to ```LinkMem``` to avoid the extra stack frames and thus I managed to get the same speed as the previous solution.

Thanks for your help! Much appreciated!

Regards,
Cristian
